### ISSUE TITLE
Colors lost when piping

### ISSUE THREAD

**Issue by justinhelmer** [mochajs/mocha#2064]
Theres a possibility I'm doing something wrong, but a little bit of insight could go a long way.

When building a custom reporter, ANSI character codes are stripped when spawning `mocha` and piping the child process `stdout` using `Base.color`. This is the undesired behavior.

Some assertions I can make to help narrow the issue:
- I am passing the `--color` option to `mocha`. I can clearly see the `--color` option working, because base reporters (tested `dot` & `spec`) correctly output color, _only_ when this switch is set.
- I am using `Base.color` "correctly," because the reporter outputs colors just fine when _not_ spawning a child process.
- I notice internally, `Base.color` uses [chalk](https://github.com/chalk/chalk), which in-turn uses [supports-color](https://github.com/chalk/supports-color). I tried enforcing the `FORCE_COLOR` environment variable ([used by supports-color](https://github.com/chalk/supports-color#info)), with no luck. I also validated that `FORCE_COLOR` works as expected with other (non-`mocha`) modules I have written.
- I have validated that [exports.useColors](https://github.com/mochajs/mocha/blob/119291449cd03a11cdeda9e37cf718a69a012896/lib/reporters/base.js#L42) is correctly set to `true` in the conditions under test.

When using `chalk` directly, there are no issues; the color outputs correctly in the spawned child process. However, to take advantage of `self.epilogue()` and mocha's built-in color map, it would be ideal if we could get this resolved.

**Comment by boneskull** [comment]
@justinhelmer if you can track down where exactly this is happening, please let me know

**Comment by jbnicolai** [comment]
@justinhelmer @boneskull The issue seems to be that we're currently using `chalk@0.4.0` for some reason (I wonder why. Do you know, @boneskull?).

This version does not have `FORCE_COLLOR` support yet. I'll whip up a PR updating asap

See https://github.com/mochajs/mocha/issues/1844 for reference.

**Comment by boneskull** [comment]
@jbnicolai we can't upgrade `chalk` past 0.4.0 because it uses tildes in `package.json`

**Comment by boneskull** [comment]
that still doesn't explain why we're not getting colors w/o `FORCE_COLOR`

**Comment by jbnicolai** [comment]
@boneskull do you mean carets (`^`)? `chalk@1.1.1` does not use any tildes, see: https://github.com/chalk/chalk/blob/v1.1.1/package.json

**Comment by jbnicolai** [comment]
Just to be sure, I've also recursively checked al subdebs of `chalk@1.1.1`, but there's definitely no tilde-notation anywhere.

```console
find node_modules/chalk/ -name package.json -exec grep '~' {} \;
```

**Comment by jbnicolai** [comment]
Nevermind. Doing some Googling, it seams you indeed meant carets. I'll leave the above for context.

**Comment by boneskull** [comment]
sorry, yes, carets.

**Comment by justinhelmer** [comment]
Thanks for digging into this guys.

Let me know if there's anything I can do to help.

**Comment by justinhelmer** [comment]
@boneskull -

Sorry for taking more of your time, but I couldn't locate the context relating to why upgrading the the versions of `npm` (`> 1.3.7`) or `node` (`> 0.8.x`) is considered a breaking change to the `2.x` branch of `mocha`.

I looked through #1969 and #1844 and the rest of the (open) issue queue but couldn't find it.

Is it something I might be able to fix in order to get `mocha` onto `npm 3.x`? If so, can you point me in the right direction? 

FYI the box I'm working on right now has `node 0.12.x` and `npm 3.5.x`.

**Comment by justinhelmer** [comment]
I'll also dig a bit more into your comment:
> that still doesn't explain why we're not getting colors w/o `FORCE_COLOR`

when I have some time.

**Comment by boneskull** [comment]
@justinhelmer I think you misunderstand.  We want to retain compatibility with a vanilla install of Node.js 0.8x.  This installs an npm version which has no support for carets.  This means we cannot upgrade chalk any further than v0.4.0, because it uses carets in its `package.json`.

Mocha works fine with npm 3.x, to my knowledge.

**Comment by justinhelmer** [comment]
@boneskull makes perfect sense. Thank you for clarifying.

**Comment by justinhelmer** [comment]
@boneskull 

Had a chance to actually dig in; I was able to identify the "issue" and can articulate it accurately.

Our confusion comes from the fact that I originally misidentified the issue. The **first** and **fourth** assertions from the [original post](https://github.com/mochajs/mocha/issues/2064#issue-128412709) are inaccurate. They would be correctly updated as follows:

- I am **NOT** passing the `--color` option to `mocha`. For that reason, the child process identifies `stdout` as non-`TTY` and strips ANSI character sequences.
- I have validated that exports.useColors is **NOT** correctly set to `true` in the conditions under test.

I was getting tripped up on the version mismatches. As I'm sure you know, `npm 3.x` changed to a [flat structure](https://github.com/npm/npm/blob/master/CHANGELOG.md#flat-flat-flat), and tossed `mocha` somewhere different than I was expecting (old dog, new tricks). The _parent_ process was running the _latest_ version, but the _child_ process was accessing the `Base` reporters from the **older** version of `mocha`. The two versions of `mocha` being used in a single run threw me off.

On top of that, the third party reporter I was using had an old (`1.x` branch) version of `mocha`, that didn't correctly implement color support (`supports-color` wasn't yet a part of the `Base` reporter). Passing the `--color` switch was working but `FORCE_COLOR` was not, which further led to my misidentification of the problem.

However, the observation still exists in the context you have defined it. You have correctly identified the issue as it relates to `FORCE_COLOR`; a newer version of `supports-color` would "fix" that. Arguably `mocha` has no interest in supporting this flag; perhaps documenting this limitation would be sufficient.

So the more interesting point of why `FORCE_COLOR` is needed in the first place is: **IT'S NOT.**

If there is a place where you document these subtleties, I would be happy to submit a pull request to bring these somewhat complicated, subtle road bumps to light. Unless you are not interested in categorizing these discoveries as such.

I apologize for taking your valuable time.
